ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectFormat.java

```diff
@@ -0,0 +1,142 @@
+import io.confluent.kafka.schemaregistry.ParsedSchema;
+import io.confluent.ksql.name.ColumnName;
+import io.confluent.ksql.schema.ksql.SchemaConverters;
+import io.confluent.ksql.schema.ksql.SimpleColumn;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.serde.Format;
+import io.confluent.ksql.util.DecimalUtil;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+
+public abstract class ConnectFormat implements Format {
+
+  private final Function<Schema, Schema> toKsqlTransformer;
+
+  public ConnectFormat() {
+    this(new ConnectSchemaTranslator()::toKsqlSchema);
+  }
+
+  @VisibleForTesting
+  ConnectFormat(final Function<Schema, Schema> toKsqlTransformer) {
+    this.toKsqlTransformer = Objects.requireNonNull(toKsqlTransformer, "toKsqlTransformer");
+  }
+
+  @Override
+  public boolean supportsSchemaInference() {
+    return true;
+  }
+
+  @Override
+  public List<SimpleColumn> toColumns(final ParsedSchema schema) {
+    final Schema connectSchema = toKsqlTransformer.apply(toConnectSchema(schema));
+
+    return connectSchema.fields().stream()
+        .map(ConnectFormat::toColumn)
+        .collect(Collectors.toList());
+  }
+
+  public ParsedSchema toParsedSchema(final List<SimpleColumn> columns) {
+    final SchemaBuilder schemaBuilder = SchemaBuilder.struct();
+    columns.forEach(col -> schemaBuilder.field(
+        col.ref().name(),
+        SchemaConverters.sqlToConnectConverter().toConnectSchema(col.type()))
+    );
+
+    final Schema connectSchema = ensureNamed(schemaBuilder.build());
+    return fromConnectSchema(connectSchema);
+  }
+
+  protected abstract Schema toConnectSchema(ParsedSchema schema);
+
+  protected abstract ParsedSchema fromConnectSchema(Schema schema);
+
+  private static SimpleColumn toColumn(final Field field) {
+    final ColumnName name = ColumnName.of(field.name());
+    final SqlType type = SchemaConverters.connectToSqlConverter().toSqlType(field.schema());
+    return new ConnectColumn(name, type);
+  }
+
+  private static Schema ensureNamed(final Schema schema) {
+    final SchemaBuilder builder;
+    switch (schema.type()) {
+      case BYTES:
+        DecimalUtil.requireDecimal(schema);
+        builder = DecimalUtil.builder(schema);
+        break;
+      case ARRAY:
+        builder = SchemaBuilder.array(ensureNamed(schema.valueSchema()));
+        break;
+      case MAP:
+        builder = SchemaBuilder.map(
+            Schema.STRING_SCHEMA,
+            ensureNamed(schema.valueSchema())
+        );
+        break;
+      case STRUCT:
+        builder = SchemaBuilder.struct();
+        if (schema.name() == null) {
+          builder.name("TestSchema" + UUID.randomUUID().toString().replace("-", ""));
```

Comment:
this seems like it doesn't belong in prod code?

Reply:
I agree. However, I've not added it, only moved it.  It was in `TestCaseBuilderUtil.addNames()`.

It's a bit of a pain, because KSQL types don't support naming, only Connect Schemas do. However, some formats, (PB & Avro), require structs to be named.

admittedly, this was a bit lazy of me. Let me take a look into doing it a better way.

Reply:
Found a way to remove this.